numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -624,6 +682,12 @@ PyArray_ConcatenateInto(PyObject *op, int axis, PyArrayObject *ret)
                         "The first input argument needs to be a sequence");
         return NULL;
     }
+    if (ret != NULL && dtype != NULL) {
+        PyErr_SetString(PyExc_TypeError,
+                "concatenate() only takes `out` or `dtype` as an "
+                "argument, but both were provided.");
+        return NULL;
+    }
```

Comment:
`ufunc`s don't seem to share this logic, and it would be nice if they could be consistent.

Do you know what ufuncs are doing when passed both arguments?

Reply:
True, but do you feel its  useful? Ufuncs will use the given dtype for ouput dtypes (should usually be all inputs instead). So basically, the use the dtype to figure out which dtype to do the calculation. They will then happily cast and write the reuslt into `out` if that is "same_kind" possible.

So yes, you could allow that, but that would potentially mean two cast steps and unlike ufuncs there is no computation for which the dtype selection is differnent from the output dtype selection.

I suppose I am simply not convinced it is actually useful to pass both, to me it would be more like adding a `dtype` argument to `take` and then supporting it mixed with `out`, and that seems strange/unnecessary as well.

For ufuncs I guess its also a bit strange, but at least there you could say it makes sense to do: `float32, float32 -> int64` (arrays), but do the calculation as `float64, float64 -> float64` (with `casting="unsafe"`, because you want to do a cast, but only after you did a higher precision float operation. To be honest, it all seems far fetched.

Reply:
Sorry, should have deleted the first chunk of that reply... In any case, the `dtype` for ufuncs to me means more: The dtype in which the computation is performed. Since there is no computation performed here, I am not sure the similarity is actually all that clear.

Reply:
Thanks, I was just asking for clarification on the difference - I agree that there's little point supporting both here.